src/main/java/com/exalttech/trex/ui/controllers/dashboard/filters/DashboardStreamsSelector.java

```diff
@@ -0,0 +1,165 @@
+
+    private ActivePGIDsService activePGIDsService = new ActivePGIDsService();
+    private Set<Integer> pgids = new HashSet<>();
+    private Map<Integer, String> selectedPGIds = null;
+    private EventHandler<Event> onFiltersChanged;
+
+    public DashboardStreamsSelector() {
+        Initialization.initializeFXML(
+                this,
+                "/fxml/Dashboard/filters/DashboardStreamsSelector.fxml"
+        );
+
+        Initialization.initializeCloseEvent(root, this::onWindowCloseRequest);
+
+        activePGIDsService.setPeriod(Duration.seconds(Constants.REFRESH_ONE_INTERVAL_SECONDS));
+        activePGIDsService.setOnSucceeded(this::handleStreamsReceived);
+        activePGIDsService.start();
+    }
+
+    public EventHandler<Event> getOnFiltersChanged() {
+        return onFiltersChanged;
+    }
+
+    public void setOnFiltersChanged(final EventHandler<Event> onFiltersChanged) {
+        this.onFiltersChanged = onFiltersChanged;
+    }
+
+    public Map<Integer, String> getSelectedPGIds() {
+        return selectedPGIds != null ? selectedPGIds : new HashMap<>();
+    }
+
+    private void handleStreamsReceived(final WorkerStateEvent event) {
+        final Set<Integer> receivedPGIds = (Set<Integer>) event.getSource().getValue();
+
+        if (receivedPGIds == null) {
+            return;
+        }
+
+        synchronized (receivedPGIds) {
+            if (pgids.equals(receivedPGIds)) {
+                return;
+            }
+
+            pgids = new HashSet<>(receivedPGIds);
+        }
+
+        if (!pgids.isEmpty() && selectedPGIds == null) {
+            selectedPGIds = new HashMap<>();
+            int count = 4;
+            for (final int pgid : pgids) {
+                selectedPGIds.put(pgid, holdColor());
+                if (--count == 0) {
+                    break;
+                }
+            }
+            handleFiltersUpdated();
+        }
+
+        build();
+    }
+
+    private void onWindowCloseRequest(final WindowEvent window) {
+        if (activePGIDsService.isRunning()) {
+            activePGIDsService.cancel();
+        }
+    }
+
+    private void build() {
+        selectedStreamsContainer.getChildren().clear();
+        streamsContainer.getChildren().clear();
+
+        int index = 0;
+        for (final Map.Entry<Integer, String> entry : selectedPGIds.entrySet()) {
+            final Integer pgid = entry.getKey();
+            selectedStreamsContainer.getChildren().add(
+                    new DashboardSelectedStream(
+                            pgid,
+                            entry.getValue(),
+                            !pgids.contains(pgid),
+                            this::handleStreamDeleteClicked
+                    )
+            );
+        }
+
+        for (final int pgid : pgids) {
+            if (selectedPGIds.get(pgid) == null) {
+                streamsContainer.getChildren().add(new DashboardStream(pgid, this::handleStreamAddClicked));
+            }
+        }
```

Comment:
No, it at least the same or even better :) It doesn't traverse twice. I don't insist on rewriting it with stream API. Just wanted to let you know.